pygments/lexers/ul4.py

```diff
@@ -0,0 +1,261 @@
+"""
+    pygments.lexers.ul4
+    ~~~~~~~~~~~~~~~~~~~
+
+    Lexer for the UL4 templating language.
+
+    More information: https://python.livinglogic.de/UL4.html
+
+    :copyright: Copyright 2006-2022 by the Pygments team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+"""
+
+import re
+
+from pygments import lexer, token
+from pygments.lexers import web, python
+
+__all__ = [
+    'UL4Lexer',
+    'HTMLUL4Lexer',
+    'XMLUL4Lexer',
+    'CSSUL4Lexer',
+    'JavascriptUL4Lexer',
+    'PythonUL4Lexer',
+]
+
+
+class UL4Lexer(lexer.RegexLexer):
+    """
+    Generic lexer for UL4.
+
+    .. versionadded:: 2.12
+    """
+
+    flags = re.MULTILINE | re.DOTALL | re.UNICODE
+
+    name = 'UL4'
+    aliases = ['ul4']
+    filenames = ['*.ul4']
+
+    tokens = {
+        "root": [
+            (
+                # Template header without name:
+                # ``<?ul4?>``
+                r"(<\?)(\s*)(ul4)(\s*)(\?>)",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword, token.Text, token.Comment.Preproc),
+            ),
+            (
+                # Template header with name (potentially followed by the signature):
+                # ``<?ul4 foo(bar=42)?>``
+                r"(<\?)(\s*)(ul4)(\s*)([a-zA-Z_][a-zA-Z_0-9]*)?\b",
+                lexer.bygroups(token.Comment.Preproc, token.Text, token.Keyword, token.Text, token.Token.Name.Function),
```

Comment:
Please use Whitespace here and elsewhere (https://github.com/pygments/pygments/issues/1905).

Reply:
Fixed in https://github.com/doerwalter/pygments/commit/69365c5427ac9ac88f5a5f9352fed0d789ae0564

Reply:
... and in https://github.com/doerwalter/pygments/commit/ef5d9029bedef4454e0f39ad682c0aaee5e0061d

Reply:
... and in https://github.com/doerwalter/pygments/commit/0a3a687b62aa258e2cd2425b07a705f9eac8ff0e